### PR TITLE
travel statistics #181414773

### PR DESCRIPTION
**Pull request by itisWasp**
## 1. Name of task
Users should get stats of trips made in the last X timeframe
## 2. Pivot Tracker ID
https://www.pivotaltracker.com/story/show/181414773
## 3. Branch Name
ft-travel-statistics-181414618
## 4. Description of task
Why
As an authenticated user, I should get the stats in terms of number of trips made in the last months, days, weeks etc.
Managers and travelers should get detailed statistics of the number of trips he/she has made in the recent past

## 7. Logic on the task to be completed
-The user should provide the period that he/she wants the trips he has made and the system should calculate the approved trips in the range provided by the user and the is of the user
-Manager should also provide the period of time that he/she wants the trips that he has managed the the system should calculate the trips in that range.

## 8. Logic On Actual Implementation
This was implemented by creating two GET end points. The first one is about get the statistics of trips made in specific time range by the user or manager and the second one is about getting the trips made in x time back.

## 9. How can this be manually tested
-Swagger: After reaching swagger doc You first login as requester or as manager Then you go under the trip tag on the GET end point of statistics . There are two endpoints the first one you put in the box the time range that you want the trips statistics then you execute (N.B the date must be valid and less than today's date and the end date must be greater than starting date). The second you put the period which can be day, week, month or year(you can but them in plural and it is case insensitive).then you put the number of period. Both end point if successfully give out 200 status code or 404 if not or 500 for server error.

### CODE REVIEW

**Review by Robertishimwe — APPROVED** [review]

**Review by niyojuste — APPROVED** [review]

**Review by sndicunguye — APPROVED** [review]

File: src/controllers/tripStatistics.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
+import { countTrips } from '../services/tripService';
+import successResponse from '../utils/successResponse';
+
+export const tripStatistics = async (req, res) => {
+  const userId = req.user.id;
+  const { start } = req.query;
+  let { end } = req.query;
+  const endingDate = end;
+  const isToday =
+    new Date(end).toLocaleDateString() === new Date().toLocaleDateString();
```
Comment (didasmbalanya) [inline src/controllers/tripStatistics.js:10]: Please re check this end date. if the format changes it wont work

`
const today = new Date().toLocaleDateString() // 5/18/2022
const end = new Date(18/5/2022).toLocaleDateString() // '1/1/1970'
`

comparing this two will fail

File: src/controllers/tripStatistics.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    end = new Date(end);
+    end.setDate(end.getDate() + 1);
+  }
+
+  const userRole = req.roleId;
+  if (userRole !== 1 && userRole !== 2) {
+    return res
+      .status(403)
+      .json('Statistics of travel are available for requester and Manager');
+  }
```
Comment (didasmbalanya) [inline src/controllers/tripStatistics.js:23]: move this to a middleware

File: src/validations/tripValidation.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
+import DateExtension from '@joi/date';
+import * as JoiImport from 'joi';
+
+const Joi = JoiImport.extend(DateExtension);
+
+const staticsValidate = Joi.object().keys({
+  start: Joi.date()
+    .utc('javascript')
```
Comment (didasmbalanya) [inline src/validations/tripValidation.js:8]: This is what I mean, UTC allows different formats. eg DD-MM-YYYY. which is what failed in my PR comment above

**Review by didasmbalanya — CHANGES_REQUESTED** [review]

File: src/services/tripService.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
+/* eslint-disable import/prefer-default-export */
+import { Op } from 'sequelize';
+import models from '../database/models';
+
+const { TripRequest, Accommodation, Users } = models;
+
+export const countTrips = async (userId, start, end, userRole) => {
```
Comment (didasmbalanya) [inline src/services/tripService.js:7]: what about other user roles?

**Review by didasmbalanya — CHANGES_REQUESTED** [review]

**Pull request merged**